File: h5gl/src/Pan3d/mateial/MaterialParam.ts
module Pan3d {
    export class MaterialParam  extends ResCount {
        public material: Material;
        public materialUrl: string;

        // public program: WebGLProgram;
        public shader: Shader3D;
        public dynamicTexList:Array<DynamicTexItem>;
        public dynamicConstList:Array<DynamicConstItem>;

    
    

        public setMaterial($materialTree: Material): void {
           
            this.material = $materialTree;
            this.materialUrl = $materialTree.url;

            this.dynamicTexList = new Array;
            this.dynamicConstList = new Array;
            this.setTexList();
            this.setConstList();
           
        }

        public setLife($life: number): void {
      
            for (var i: number = 0; i < this.dynamicTexList.length; i++) {
                if (this.dynamicTexList[i].isParticleColor) {
                    this.dynamicTexList[i].life = $life;
                }
            }
            
        }



        public setTexList(): void {
         
            var texList: Array<TexItem> = this.material.texList;
            for (var i: number = 0; i < texList.length; i++) {
                var dyTex: DynamicTexItem;
                if (texList[i].isParticleColor) {
                    dyTex = new DynamicTexItem(this.scene3D);
                    dyTex.target = texList[i];
                    dyTex.paramName = texList[i].paramName;
                    dyTex.initCurve(4);
                    this.dynamicTexList.push(dyTex);
                    dyTex.isParticleColor = true;
                } else if (texList[i].isDynamic) {
                    dyTex = new DynamicTexItem(this.scene3D);
                    dyTex.target = texList[i];
                    dyTex.paramName = texList[i].paramName;
                    this.dynamicTexList.push(dyTex);
                }

            }
          
        }

        public setConstList(): void {
           
            var constList: Array<ConstItem> = this.material.constList;

            for (var i: number = 0; i < constList.length; i++) {
                var constItem: ConstItem = constList[i];
                var dyCon: DynamicConstItem;
                if (constItem.param0Type != 0) {
                    dyCon = new DynamicConstItem(this.scene3D);
                
                    dyCon.setTargetInfo(constItem, constItem.paramName0, constItem.param0Type);
                    this.dynamicConstList.push(dyCon);
                }

                if (constItem.param1Type != 0) {
                    dyCon = new DynamicConstItem(this.scene3D);
     
                    dyCon.setTargetInfo(constItem, constItem.paramName1, constItem.param1Type);
                    this.dynamicConstList.push(dyCon);
                }

                if (constItem.param2Type != 0) {
                    dyCon = new DynamicConstItem(this.scene3D);
       
                    dyCon.setTargetInfo(constItem, constItem.paramName2, constItem.param2Type);
                    this.dynamicConstList.push(dyCon);
                }

                if (constItem.param3Type != 0) {
                    dyCon = new DynamicConstItem(this.scene3D);
               
                    dyCon.setTargetInfo(constItem, constItem.paramName3, constItem.param3Type);
                    this.dynamicConstList.push(dyCon);
                }
            }
 
        }

        public setTextObj(ary: Array<any>): void {
        
            for (var i: number = 0; i < ary.length; i++) {
                var obj: any = ary[i];
                for (var j: number = 0; j < this.dynamicTexList.length; j++) {
                    var dynamicTexItem:DynamicTexItem=this.dynamicTexList[j];
                    if (dynamicTexItem.paramName == obj.paramName) {
                        if (dynamicTexItem.isParticleColor) {
                            dynamicTexItem.curve.setData(obj.curve);
                        } else {
                            dynamicTexItem.url = obj.url;
                        }
                        break;
                    }
                }
            }
        

        }

        public setConstObj(ary: Array<any>): void {
         
            for (var i: number = 0; i < ary.length; i++) {
                var obj: any = ary[i];
                for (var j: number = 0; j < this.dynamicConstList.length; j++) {
                    if (this.dynamicConstList[j].paramName == obj.paramName) {
                        this.dynamicConstList[j].curve.setData(obj.curve)
                        break;
                    }
                }
            }
          
        }




    }
}

File: android/android/app/src/main/java/com/pan3d/material/MaterialParam.java
package com.pan3d.material;

import com.pan3d.program.Shader3D;
import com.pan3d.scene.Scene3D;
import com.pan3d.vo.ParamDataConVo;
import com.pan3d.vo.ParamDataVo;

import java.util.ArrayList;
import java.util.List;

public class MaterialParam {

    private static final String TAG = "MaterialParam";
    public Material material ;
    public String materialUrl  ;
    public Shader3D shader3D;
    public List<DynamicTexItem> dynamicTexList;
    public List<DynamicConstItem>  dynamicConstList ;
public Scene3D scene3D;
    public MaterialParam(Scene3D val){
        scene3D=val;
    }



    public void setMaterial(Material val) {
        this.material = val;
        this.materialUrl =val.url;
        this.dynamicTexList = new ArrayList<>();
        this.dynamicConstList =new ArrayList<>();
        this.setTexList();
        this.setConstList();
    }

    private void setConstList() {
        List<ConstItem> constList = this.material.constList;
        for (int i = 0; i < constList.size(); i++) {
            ConstItem constItem  = constList.get(i);
            DynamicConstItem dyCon ;
            if (constItem.param0Type != 0) {
                dyCon = new DynamicConstItem();
                dyCon.setTargetInfo(constItem, constItem.paramName0, constItem.param0Type);
                this.dynamicConstList.add(dyCon);
            }
            if (constItem.param1Type != 0) {
                dyCon = new DynamicConstItem();
                dyCon.setTargetInfo(constItem, constItem.paramName1, constItem.param1Type);
                this.dynamicConstList.add(dyCon);
            }
            if (constItem.param2Type != 0) {
                dyCon = new DynamicConstItem();
                dyCon.setTargetInfo(constItem, constItem.paramName2, constItem.param2Type);
                this.dynamicConstList.add(dyCon);
            }
            if (constItem.param3Type != 0) {
                dyCon = new DynamicConstItem();

                dyCon.setTargetInfo(constItem, constItem.paramName3, constItem.param3Type);
                this.dynamicConstList.add(dyCon);
            }
        }
    }

    protected void setTexList() {
        List<TexItem> texList  = this.material.texList;
        for (int i = 0; i < texList.size(); i++) {
            DynamicTexItem dyTex ;
            TexItem texItem= texList.get(i);
            if (texItem.isParticleColor) {
                dyTex = new DynamicTexItem(scene3D);
                dyTex.target = texItem;
                dyTex.paramName =texItem.paramName;
                dyTex.initCurve(4);
                this.dynamicTexList.add(dyTex);
                dyTex.isParticleColor = true;
            } else if (texItem.isDynamic) {
                dyTex = new DynamicTexItem(scene3D);
                dyTex.target = texItem;
                dyTex.paramName = texItem.paramName;
                this.dynamicTexList.add(dyTex);
            }

        }
    }

    public void setLife(float life) {
        for (int i = 0; i < this.dynamicTexList.size(); i++) {
            if (this.dynamicTexList.get(i).isParticleColor) {
                this.dynamicTexList.get(i).life =life;
            }
        }
    }

    public void setTextObj(List<ParamDataVo> ary) {
        for (int i = 0; i < ary.size(); i++) {
            ParamDataVo obj =(ParamDataVo) ary.get(i);
            for (int j = 0; j < this.dynamicTexList.size(); j++) {
                DynamicTexItem dynamicTexItem=    this.dynamicTexList.get(j);
                if (dynamicTexItem.paramName.equals(obj.paramName)) {
                    if (dynamicTexItem.isParticleColor) {
                        dynamicTexItem.curve.setData(obj.curve);
                    } else {
                        dynamicTexItem.url = obj.url;
                    }
                    break;
                }
            }
        }

    }

    public void setConstObj(List<ParamDataConVo> ary) {
        for (int i = 0; i < ary.size(); i++) {
            ParamDataConVo obj = ary.get(i);
            for (int j = 0; j < this.dynamicConstList.size(); j++) {
                DynamicConstItem dynamicConstItem=    this.dynamicConstList.get(j);
                if (dynamicConstItem.paramName.equals(obj.paramName)) {
                    dynamicConstItem.curve.setData(obj.curve);
                    break;
                }
            }
        }


    }

}
